### ISSUE TITLE
Inquiries Regarding AaD and SHOT Algorithms

### ISSUE THREAD

**Issue by yichenwang231** [mohamedr002/MAPU_SFDA_TS#3]
AaD: the two queues used in the adaptation process are randomly initialized at each iteration, only the data of the current batch is deposited into the queue, then the negative examples selected are randomly generated, which is inconsistent with the AaD source code and the paper
SHOT: used feature_extractor.eval() when computing pseudo-labels by clustering, but did not return feature_extractor to training mode after clustering was finished, also, indexing pseudo-labels of the current batch by trg_idx during iteration is wrong, trg_idx and the obtained pseudo_labels don't have such a correspondence, I don't know if this is a discrepancy due to hardware issues, I am anticipating your answer. Thanks.

**Comment by mohamedr002** [comment]
Thanks for bringing up this issue. Regarding the AaD, I've actually resolved the problem, but it appears the fix hasn't been pushed yet. You're correct that the current feature and score banks are random. We should indeed build these banks from the data before using them. Below is the function I used for this purpose. I will ensure the updates are pushed shortly:


```
def build_feat_score_bank(self, data_loader):
        fea_bank = torch.empty(0).cuda()
        score_bank = torch.empty(0).cuda()

        self.feature_extractor.eval()
        self.classifier.eval()

        # Process data batch by batch
        for data in data_loader:
            batch_data = data[0].cuda()  # Assuming the first element in your batch is the data. Adjust as needed.

            batch_feat,_ = self.feature_extractor(batch_data)
            norm_feats = F.normalize(batch_feat)
            batch_pred = self.classifier(batch_feat)
            batch_probs = nn.Softmax(dim=-1)(batch_pred)

            # Update the banks
            fea_bank = torch.cat((fea_bank, norm_feats.detach()), 0)
            score_bank = torch.cat((score_bank, batch_probs.detach()), 0)

        return fea_bank, score_bank

```

**Comment by mohamedr002** [comment]
Regarding SHOT Method: 
- With regard to eval, I have also fixed this issue after uploading the code in my local and it seems that the updates pushed to another repo, by adding 
```
  self.feature_extractor.train()
  self.classifier.train()
```
At the end of the pseudo label function, no performance difference was noticed. 

- Regarding the correspondence between pseudo labels and trg_idx, it has already been adopted by SHOT methods authors. Besides, I disagree with you that there is no correspondence. Actually, I make sure that shuffling is false to ensure that trg_idx corresponds to the predicted label.

**Comment by yichenwang231** [comment]
> * Actually, I make sure that shuffling is false to ensure that trg_idx corresponds to the predicted label.

I had overlooked the shuffle setting, thank you very much!
